kernel/src/process_loading.rs

```diff
@@ -0,0 +1,769 @@
+            app_memory.as_ptr() as usize,
+            app_memory.as_ptr() as usize + app_memory.len() - 1
+        );
+    }
+
+    let mut remaining_flash = app_flash;
+    let mut remaining_memory = app_memory;
+    // Try to discover up to `procs.len()` processes in flash.
+    let mut index = 0;
+    let num_procs = procs.len();
+    while index < num_procs {
+        let load_result = load_process(
+            kernel,
+            chip,
+            remaining_flash,
+            remaining_memory,
+            index,
+            fault_policy,
+            capability,
+        );
+        match load_result {
+            Ok((new_flash, new_mem, proc)) => {
+                remaining_flash = new_flash;
+                remaining_memory = new_mem;
+                if proc.is_some() {
+                    if config::CONFIG.debug_load_processes {
+                        proc.map(|p| debug!("Loaded process {}", p.get_process_name()));
+                    }
+                    procs[index] = proc;
+                    index = index + 1;
+                }
+            }
+            Err((_new_flash, _new_mem, err)) => {
+                if config::CONFIG.debug_load_processes {
+                    debug!("No more processes to load: {:?}.", err);
+                }
+                // No more processes to load.
+                break;
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Use `checker` to transition `procs` from the `Unverified` into the
+/// `Unstarted` state (if they pass the checker policy) or
+/// `CredentialsFailed` state (if they do not pass the checker
+/// policy). If `verifier` is `None` then all processes are
+/// automatically verified. Processes that transition into the
+/// `Unstarted` state are started by enqueueing a stack frame to run
+/// the initialization function as indicated in the TBF header.
+#[inline(always)]
+fn check_processes<'a, KR: KernelResources<C>, C: Chip>(
+    procs: &'static [Option<&'static dyn Process>],
+    kernel_resources: &KR,
+) -> Result<(), ProcessLoadError>
+//    where <KR as KernelResources<C>>::CredentialsCheckingPolicy: 'static,
+{
+    let policy = kernel_resources.credentials_checking_policy();
+    #[allow(unused_mut)]
+    let machine = unsafe {
+        static_init!(
+            ProcessCheckerMachine,
+            ProcessCheckerMachine {
+                process: Cell::new(0),
+                footer: Cell::new(0),
+                checker: OptionalCell::empty(),
+                processes: procs,
+            }
+        )
+    };
+    policy.set_client(machine);
+    machine.checker.replace(policy);
+    machine.next()?;
+    Ok(())
+}
+
+/// Iterates across the `processes` array, checking footers and deciding
+/// whether to make them runnable based on the checking policy in `checker`.
+/// Starts processes that pass the policy and puts processes that don't
+/// pass the policy into the `CredentialsFailed` state.
+struct ProcessCheckerMachine {
```

Comment:
Should this be in `process_checking.rs` rather than `process_loading.rs`?

Reply:
It should be here: `process_checking` defines the APIs, and `process_loading` has one implementation of the API. One could have a different process loader that has a different implementation.